src/lake/Lake/Build/Trace.lean

```diff
@@ -137,10 +137,38 @@ instance : ComputeHash FilePath IO := ⟨computeFileHash⟩
 structure TextFilePath where
   path : FilePath
 
+/-- This is the same as `String.replace text "\r\n" "\n"`, but more efficient. -/
+partial def replaceCRLFToLF (text : String) : String :=
+  start 0
+where
+  start (pos : String.Pos) : String :=
```

Comment:
Just wondering, did you measure the impact of this duplication? I assume the final append is insignificant

Reply:
It's less about the append than the fact that a new allocation is needed. The case where there are no `\r\n` in the file is common (e.g. all files on linux/mac) and so it is important to have a fast path for this case.

Reply:
The empty string allocation by `extract`?

Reply:
The effect is small (about 5-10%) but measurable. Adding `if accStop = 0 then text else acc ++ ...` into the base case of `go` avoids the duplication but is about as fast as just calling `go "" 0 0` directly and lacks the 5% boost mentioned. (Note: benchmarks were run on the interpreted code, I did not try the compiled version.)

Reply:
It doesn't really make sense to benchmark the interpreted code if the goal is to have the code be compiled. Even just the overhead of passing around the two additional arguments is likely incomparable between interpreted and compiled.

Reply:
> The empty string allocation by extract?

no, in the case where there is nothing to replace in the string the returned result is effectively `"" ++ text.extract 0 text.endPos` which is a full copy of `text`. I checked, it does not produce the same (pointer-equal) string, and I don't see a fast path for this in either `lean_string_utf8_extract` or `lean_string_append` so I think it might even make two copies of `text`.

Reply:
Okay, I did the compiled benchmark, and it looks even more competitive, 9x faster rather than 4x:
```
new 46.9ms
41590784
old 403ms
41590784
```
Calling `go "" 0 0` instead of `start 0` results in `101ms` instead so the difference is much more easily measurable here. Adding the `if` statement to `go` makes it go back down to about `46ms` +- 2ms so I'll use that version instead (which avoids having to copy the main logic).